### PR TITLE
ENH: Make numpy.array_api more portable

### LINKED ISSUE #23883: ENH: Make the dtype objects in numpy.array_api more strict
### Proposed new feature or change:

In [NEP 47](https://numpy.org/neps/nep-0047-array-api-standard.html#the-dtype-objects), we decided to reuse normal `numpy` dtype objects in `numpy.array_api`. However, as we've come to use the array API standard more and learn about how it is used, we've learned that `numpy.array_api` is primarily useful as a strict namespace for libraries to test against to ensure they don't deviate from the standard. We've also learned that it's very difficult to mix `numpy.array_api` with normal `numpy`, and usage of actual `numpy` `ndarray`s should take a [different approach](https://github.com/data-apis/array-api-compat).

Therefore, I think it makes more sense to use separate dtype objects in numpy.array_api, similar to how we use a distinct `Array` object. This would remove all dtype attributes and behaviors from the NumPy dtype objects that aren't guaranteed by the standard. That is, [nothing should be implemented on the objects except for `__eq__`](https://data-apis.org/array-api/2022.12/API_specification/data_types.html#data-type-objects), and `__eq__` itself should only compare directly against the objects themselves, not against things like dtype strings. 

Note that this isn't particularly high priority, as currently any library that uses NumPy dtype specific behaviors will also detect this as soon as they test against PyTorch, whose dtype objects don't share anything in common with NumPy. But it is still a good idea in my opinion and fits with the overall goal of the module. With that being said, I don't plan to implement this for the 1.25 release. 

If someone else wants to implement this, it should be straightforward:

- Replace the current [dtype objects](https://github.com/numpy/numpy/blob/main/numpy/array_api/_dtypes.py) in `numpy.array_api` with something like
  ```py
  class Dtype:
      def __init__(self, name):
          self._name = name
          self._np_dtype = np.dtype(name)
      def __repr__(self):
          return self._name
      def __eq__(self, other):
          return isinstance(other, Dtype) and self._name == other._name

   int8 = Dtype('int8')
   int16 = Dtype('int16')
   ...
   ```

- Everywhere that a dtype object is passed to an actual NumPy function in the `numpy.array_api` code replace `dtype` with `dtype._np_dtype`.

### PR DESCRIPTION
**Pull request by asmeurer**
- Make `.device` return a special `CPU_DEVICE` object (not present in the namespace) instead of `"cpu"`. See https://github.com/data-apis/array-api/issues/626.
- Make dtype objects separate classes, rather than reusing NumPy dtype objects. Fixes https://github.com/numpy/numpy/issues/23883
- Fix `nonzero` to disallow 0-D inputs (this is not related to portability, but it was a small change so I figured I would just include it here).
- Update the array API compatibility document accordingly. 

These will make it harder to write non-portable code that successfully works with `numpy.array_api`.

### CODE REVIEW

File: numpy/array_api/_dtypes.py [hunk 1]
```diff
@@ -1,21 +1,42 @@
...
+
+class _DType:
+    def __init__(self, np_dtype):
+        np_dtype = np.dtype(np_dtype)
+        self._np_dtype = np_dtype
+
+    def __repr__(self):
+        return f"np.array_api.{self._np_dtype.name}"
+
+    def __eq__(self, other):
```
Comment (asmeurer) [inline numpy/array_api/_dtypes.py:14]: I'm thinking it might be a good idea to make `==` explicitly raise an exception when compared against a `numpy` dtype object. Right now `numpy.float32 == numpy.array_api.float32` returns `False`, but I feel like this could mask bugs. Is there a straightforward way to detect if `other` is a NumPy dtype object?

**Comment by charris** [comment]
This could use a release note.

File: numpy/array_api/_typing.py [hunk 2]
```diff
@@ -50,7 +50,7 @@ class NestedSequence(Protocol[_T_co]):
     def __getitem__(self, key: int, /) -> _T_co | NestedSequence[_T_co]: ...
     def __len__(self, /) -> int: ...
 
-Device = Literal["cpu"]
+Device = type(CPU_DEVICE)
```
Comment (BvB93) [inline numpy/array_api/_typing.py:53]: Dynamically assigning a type like this isn't gona work I'm afraid. Probably want to go with an `Any` here.
```suggestion
Device = Any
```

Comment (asmeurer) [reply numpy/array_api/_typing.py:53]: We do know the type https://github.com/numpy/numpy/pull/25370/files#diff-79e6f716d807a9f60474911f90c087a717802dd6e63c82cd11689b7e47e943d1R45. I just figured this would work.

File: numpy/array_api/_dtypes.py (hunk 1 above)
Comment (mhvk) [reply numpy/array_api/_dtypes.py:14]: This is a very fly-by comment, but in python generally `==` is supposed never to fail.

Comment (asmeurer) [reply numpy/array_api/_dtypes.py:14]: Yes, I'm not completely sure about it. I did set this to error when I was creating this PR and it definitely helped me to find all the places in the code that needed to be updated. 

My worry if is someone is incorrectly doing:

```py
if x.dtype == np.float32: # x is a numpy.array_api array
    ...
```

or

```py
if x.dtype == 'f4':
    ...
```

The code will just subtly start being wrong.

But then again, I don't know if that kind of error actually exists in the wild. It is already wrong for other libraries like PyTorch.

OTOH, like you say, `==` failing can be very annoying as it would make things like `in` not work properly. I think I'm leaning towards not making this change, but I'm interested to hear what others think.

Comment (seberg) [reply numpy/array_api/_dtypes.py:14]: Notwithstanding that this should move eventually.  For *this* namespace it seems right to raise errors liberally, since I think by now it is accepted as a *testing* implementation:  If it works with this, it should work with anything.

Comment (mhvk) [reply numpy/array_api/_dtypes.py:14]: How about a warning? For testing purposes, those can be turned into errors.

Comment (seberg) [reply numpy/array_api/_dtypes.py:14]: I don't have an opinion, TBH.  Besides testing, I never saw much purpose to the namespace as is and in that context I think liberally raising is OK too.

OTOH, if you formalize the "this is for testing" part of the namespace, then it does seem quite logical to have a custom `UndefinedBehavior` warning.  That would make it easier to ignore it in tests if there is a reason to do so (not sure there is, but maybe it doesn't matter).

**Review by rgommers — COMMENTED** [review]
Thanks @asmeurer. This looks quite close, I found one issue when testing this on SciPy. Also the `==` warning/exception for dtype comparison needs to be finalized.

**Comment by rgommers** [comment]
Resolving the "2 blank lines" part of the linter complaints would also be nice to take along.

File: numpy/array_api/_creation_functions.py [hunk 3]
```diff
@@ -68,6 +63,8 @@ def asarray(
         # Give a better error message in this case. NumPy would convert this
         # to an object array. TODO: This won't handle large integers in lists.
         raise OverflowError("Integer out of bounds for array dtypes")
+    if dtype is not None:
+        dtype = dtype._np_dtype
```
[the remark this answers is not in the document]
Comment (asmeurer) [reply numpy/array_api/_creation_functions.py:67]: Good catch. The test suite didn't catch this because it failed on the NotImplementedError.

**Comment by rgommers** [comment]
The `test_warning_calls` failures are real. 

And now that gh-25317 is merged, it'd be good to update this PR for device handling of `fftfreq`/`rfftfreq`.

**Comment by rgommers** [comment]
This LGTM now, but the added strictness is hard to deal with without the fix suggested in https://github.com/data-apis/array-api-compat/issues/77#issuecomment-1894095240. So ideally that would be fixed first before merging this.

**Comment by asmeurer** [comment]
@rgommers I'm not following how that array-api-compat issue relates to this.

**Comment by rgommers** [comment]
> @rgommers I'm not following how that array-api-compat issue relates to this.

I ran into that issue when testing this PR with SciPy. All my attempts at fixing up SciPy to be robust against changes in this PR are blocked until that issue is fixed. Basically, the addition of `numpy.ndarray.__array_namespace__` changed the behavior of `array_api_compat` - but that was still okay when this module had some flexibility. The new dtype objects added in this PR are very restrictive, so it's impossible to get away with mixing `numpy` and `numpy.array_api`, or using `dtype=bool` (the Python builtin `bool`) after this gets merged.

tl;dr I don't really want to leave SciPy (and scikit-learn) in a state where tons of tests fail with NumPy `main`.

**Comment by asmeurer** [comment]
I guess there must be some complexity in the scipy implementation for this that to be related to this. I'm surprised you have array-api compatible code that's mixing numpy dtypes, since that already doesn't work for pytorch or cupy.

**Comment by rgommers** [comment]
Now that https://github.com/scipy/scipy/pull/19885 is merged, let's give this a go too. Thanks @asmeurer & reviewers!

**Comment by rgommers** [comment]
@asmeurer it looks like we're ready now for you to create the standalone version of `numpy.array_api`, so that we can make an announcement and switch over the known consumers and remove `numpy.array_api` before the 2.0.x branch point.
